File: python/Remove Zero Sum Consecutive Nodes from Linked List.py
# Definition for singly-linked list.
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeZeroSumSublists(self, head: Optional[ListNode]) -> Optional[ListNode]:
        front = ListNode(0, head)
        start = front

        while start is not None:
            prefix_sum = 0
            end = start.next

            while end is not None:
                # Add end's value to the prefix sum
                prefix_sum += end.val

                # Delete zero sum consecutive sequence
                # by setting node before sequence to node after
                if prefix_sum == 0:
                    start.next = end.next

                end = end.next

            start = start.next

        return front.next

# Create an instance of the Solution class
sol = Solution()

# Convert the list [1, 2, -3, 3, 1] to a linked list
head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(-3)
head.next.next.next = ListNode(3)
head.next.next.next.next = ListNode(1)

# Call the removeZeroSumSublists method and print the result
result = sol.removeZeroSumSublists(head)

# Print the result
while result:
    print(result.val, end=" ")
    result = result.next

File: JAVA/java/src/RemoveZeroSumConsecutiveNodesfromLinkedList.java
import java.util.HashMap;
import java.util.Map;

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class RemoveZeroSumConsecutiveNodesfromLinkedList {
    public ListNode removeZeroSumSublists(ListNode head) {
        ListNode front = new ListNode(0);
        front.next = head;

        ListNode current = front;

        int prefixSum = 0;

        Map<Integer, ListNode> prefixSumToNode = new HashMap<>();
        prefixSumToNode.put(0, front);

        while (current != null) {
            prefixSum += current.val;
            prefixSumToNode.put(prefixSum, current);
            current = current.next;
        }

        prefixSum = 0;
        current = front;

        while (current != null) {
            prefixSum += current.val;
            current.next = prefixSumToNode.get(prefixSum).next;
            current = current.next;
        }

        return front.next;
    }

    public static ListNode arrayToLinkedList(int[] arr) {
        ListNode dummy = new ListNode();
        ListNode current = dummy;
        for (int num : arr) {
            current.next = new ListNode(num);
            current = current.next;
        }
        return dummy.next;
    }

    public static void main(String[] args) {
        RemoveZeroSumConsecutiveNodesfromLinkedList ob = new RemoveZeroSumConsecutiveNodesfromLinkedList();
        ListNode head = arrayToLinkedList(new int[]{1, 2, -3, 3, 1});
        ListNode result = ob.removeZeroSumSublists(head);
        while (result != null) {
            System.out.print(result.val + " ");
            result = result.next;
        }
    }
}
